### PR TITLE
Unification of languages

### LINKED ISSUE #277: Unified languages
Might it be good to have one site that supports all the languages? With multi-editor support coming along, doing a side-by-side comparison of Rust vs C++ might be fun. Would need:

* Language dropdown on editor
* Way of tagging each compiler with its language
* Compiler list to filter on editor's language
* API considerations? Currently all compilers are named for their short name. Maybe need a `/lang/` dir in the api path?

### PR DESCRIPTION
**Pull request by RabsRincon**
This PR aims to unify every language under the same node instance. This my third try but the first that actually works! The PR won't yet be merged, but I've created it to be able to keep track of its progress.

Feel free to test the latest version (f08ceb9) at http://godbolt.org/beta/

#### Things left to do in order to be mergable
- [x] Have old links be compatible with this new version
- [x] Be able to create new links! This one should be easy 
- [x] Fix some bugs regarding compiler selection on language change. Sometimes the default one won't be chosen. (Maybe try to get the first one if there's no default one?)
- [x] Test this, a lot. (Embed being one of the things I'm not sure about how it will work)
- [x] Improve the UI. This version is just the bare minimum to get a working setup
- [x] Update docs to reflect the changes. That includes the gcc. link in the repo description
- [x] Deprecate Editor constructor for lang? Maye this one is useful for the 1st point here
- [ ] Test & update the api paths. The one the compilation uses is more or less updated, but there is room to improve it (Note that the language is invisible to the api users. We handle that internally, but they can give us a hint!)
- [x] Right now the language list is duplicated. Find a way to get only 1 version! (Hint: Options)
- [x] Test the editor-tied windows like load from source/the examples. Those need to work too
- [x] Add language support for the rest of the panels. Right now only editor and compiler know about it
- [ ] Merge https://github.com/mattgodbolt/compiler-explorer-image/pull/50
- [x] `--language X` should launch CE with lang X and only X
And more to come

## Errors that need to be fixed from testing on Beta
- [x] Page title still expects a language after the `- `
- [x] C compilers are looked for at `/usr/local/bin`
- [x] GHC times out when checking the version
- [x] Rust does not find any compiler
- [x] Haskell does not find any compiler
- [x] D does not have a default compiler
- [x] Cppx can't compile Herb's examples sometimes
- [x] Pascal does not compile
- [x] Can't find LDC 1.6.0 -> ENOENT Error

Once the unification is merged, hopefully way cooler features can come into life easier

This closes #277

### CODE REVIEW

**Comment by RabsRincon** [comment]
Sorry if this gets you into merge hell @jaredwy

**Comment by partouf** [comment]
Should lib/languages.js perhaps include pascal or is that handled in a different way?
        'pascal': {
            id: 'pascal',
            name: 'Pascal',
            monaco: 'pascal',
            extensions: ['.pas']
        },

**Comment by RabsRincon** [comment]
@partouf Oh, nice catch there! I totally missed it. I'm in the middle of merging the two language definitions so I'll fix this along with those changes.
Thanks for taking the time to take a peek over the PR 😄

**Comment by partouf** [comment]
While you're at it though... Is there any particular reason why we're not using the examples folder to also store the Defaultly loaded example?
To me it seems a bit inconsistent to do it differently.

**Comment by mattgodbolt** [comment]
I'm so excited by this! Thank you @RabsRincon. Will play about in the branch a bit.

@partouf great idea to load the examples from the example folder.

**Comment by mattgodbolt** [comment]
@RabsRincon unless I'm mistaken the "defaultCompiler" doesn't seem to be honoured. If I switch between languages I don't get a decent default (try from an Incognito mode browser window to ensure you're not getting the last compiler you picked).

**Comment by RabsRincon** [comment]
@partouf there is no reason not to, other than 
 - It was easier to put them in the language definitions
 - That would make them appear in the "Load Examples" window. But that is maybe something that we want!

**Comment by RabsRincon** [comment]
@mattgodbolt How weird indeed! https://github.com/mattgodbolt/compiler-explorer/blob/f2a75ac467bba7431cf62cf190e00e3fdd179c69/static/compiler.js#L1156 the code to honour the default compiler is there, and it works for me (Except when  no compiler is found for it and the default g++ one is set... Which I think I have to deprecate because I think I'm seeing a bug waiting to happen with that behaviour)

**Comment by mattgodbolt** [comment]
Need to think about the API structure. CC @jaredwy we discussed changing the API endpoint to be something like `/api/$lang/$compiler/compile` so then listing languages could be listed, then compilers per language or something. That could mean we could relax the constraint that compiler IDs are unique (which maybe they don't need to be anyway, but I had trouble getting the C compiler to work as well as the C++ compiler until I made them use different IDs)

**Comment by RabsRincon** [comment]
@partouf Fixed now! The source code should be loaded from `examples/LANG/default.EXT` for every language

File: app.js [hunk 1]
```diff
@@ -336,23 +332,21 @@ function retryPromise(promiseFunc, name, maxFails, retryMs) {
 }
 
 function findCompilers() {
-    var exes = compilerPropsAT(languages.toArray(), exs => {
-        return exs.split(":").filter(x => x);
-    }, "compilers", "/usr/bin/g++");
```
Comment (RabsRincon) [inline app.js]: @mattgodbolt I don't know if removing this default actually helps. Any idea?

**Comment by cjdb** [comment]
When changing languages (e.g. C++ to Rust), compiler flags that are C++-only (e.g. `-std=c++17`) persist. If you need this to be a separate issue, please let me know :)

**Comment by RabsRincon** [comment]
Thanks for the heads up @cjdb. The error you mention is fixed in 28295f7 but it has not yet reached the beta instance
:)

**Pull request merged**